### PR TITLE
Adding hash for vesting schedule

### PR DESCRIPTION
**Pull request by ilblackdragon**
By @evgenykuzyakov: 
- recompiled the contract with docker
- all integration and unit tests pass
- Updated README.
- Added a view method to get lockup `VestingInformation` to build ✌️DeFi✌️.

### Private vesting schedule

Since vesting schedule usually starts at the date of employment it allows to de-anonymize the owner of the lockup contract.
To keep the identity private, the contract allows to hash the vesting schedule with some random salt and keep store the
hash instead of the raw vesting schedule information. In case the NEAR foundation has to terminate the vesting schedule,
it will provide the raw vesting schedule and the salt, effectively revealing the vesting schedule. The contract then
will compare the hash with the internal hash and if they match proceed with the termination.

NOTE: The private vesting schedule can only be used, if the lockup release period and the lockup duration are effectively
shadowing the vesting duration. Meaning that the lockup release ends later than the vesting release and the lockup duration
ends after the vesting cliff.
Once the lockup schedule starts before the vesting schedule (e.g. employment starts after the transfers are enabled),
the vesting schedule can't be kept private anymore.

### `2.0.0`

- Changed `vesting_schedule` initialization argument to allow it to hide the vesting schedule behind a hash to keep it private.
- Added view method `get_vesting_information` to view internal vesting information.

### CODE REVIEW

File: lockup/src/getters.rs [hunk 1]
```diff
@@ -92,33 +94,36 @@ impl LockupContract {
                         0
                     };
 
-                return (std::cmp::max(
+                let unvested_amount = match &self.vesting_information {
+                    VestingInformation::VestingSchedule(vs) => self.get_unvested_amount(vs.clone()),
+                    VestingInformation::Terminating(terminating) => terminating.unvested_amount,
+                    // There is either no information because vesting is private.
```
Comment (evgenykuzyakov) [inline lockup/src/getters.rs:100]:
```suggestion
                    // Vesting is private, so we can assume the vesting started before lockup date.
```

File: lockup/src/types.rs [hunk 2]
```diff
@@ -115,13 +118,23 @@ impl VestingSchedule {
     }
 }
 
+#[derive(Serialize, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+#[serde(tag = "type", content = "value")]
+pub enum VestingScheduleOrHash {
+    Hash(String),
```
Comment (evgenykuzyakov) [inline lockup/src/types.rs:125]: Use `Base64VecU8`
```suggestion
    Hash(Base64VecU8),
```

File: lockup/src/foundation.rs [hunk 3]
```diff
@@ -8,10 +8,10 @@ impl LockupContract {
     /// Requires 25 TGas (1 * BASE_GAS)
     ///
     /// Terminates vesting schedule and locks the remaining unvested amount.
-    pub fn terminate_vesting(&mut self) {
+    pub fn terminate_vesting(&mut self, vesting_schedule: VestingSchedule, salt: Vec<u8>) {
```
Comment (evgenykuzyakov) [inline lockup/src/foundation.rs:11]: For salt use `Base64VecU8`

File: lockup/src/lib.rs [hunk 4]
```diff
@@ -145,7 +140,7 @@ pub struct LockupContract {
     pub staking_information: Option<StakingInformation>,
 
     /// The account ID that the NEAR Foundation, that has the ability to terminate vesting.
-    pub foundation_account_id: Option<AccountId>,
+    pub foundation_account_id: AccountId,
```
Comment (evgenykuzyakov) [inline lockup/src/lib.rs:143]: It's cleaner if no vesting requires that there is no `foundation_account_id`

File: lockup/src/lib.rs [hunk 5]
```diff
@@ -168,10 +163,10 @@ impl LockupContract {
     /// - `transfers_information` - the information about the transfers. Either transfers are
     ///    already enabled, then it contains the timestamp when they were enabled. Or the transfers
     ///    are currently disabled and it contains the account ID of the transfer poll contract.
-    /// - `vesting_schedule` - if present, describes the vesting schedule for employees. Vesting
-    ///    schedule affects the amount of tokens the NEAR Foundation will get in case of
+    /// - `vesting_schedule` - Hash of vesting schedule with salt.
```
Comment (evgenykuzyakov) [inline lockup/src/lib.rs:166]: Need to update the comment

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]

File: lockup/src/types.rs [hunk 6]
```diff
@@ -115,13 +118,23 @@ impl VestingSchedule {
     }
 }
 
+#[derive(Serialize, Deserialize)]
```
Comment (evgenykuzyakov) [inline lockup/src/types.rs:121]: Only need `Serialize`

File: lockup/src/lib.rs [hunk 7]
```diff
@@ -199,12 +198,6 @@ impl LockupContract {
             env::is_valid_account_id(staking_pool_whitelist_account_id.as_bytes()),
             "The staking pool whitelist account ID is invalid"
         );
-        if foundation_account_id.is_some() {
-            assert!(
-                vesting_schedule.is_some(),
-                "Foundation account can't be added without vesting schedule"
-            )
-        }
```
Comment (evgenykuzyakov) [inline lockup/src/lib.rs:207]: This check is still needed if foundation is provided, but vesting schedule is none

File: lockup/src/lib.rs [hunk 8]
```diff
@@ -850,22 +874,6 @@ mod tests {
...
-    }
-
-    #[test]
-    #[should_panic(expected = "Foundation account can't be added without vesting schedule")]
-    fn test_init_foundation_key_no_vesting_with_release() {
-        let context = basic_context();
-        testing_env!(context.clone());
-        new_contract(true, None, Some(to_nanos(YEAR).into()), true);
-    }
-
```
Comment (evgenykuzyakov) [inline lockup/src/lib.rs:868]: Keep those

File: lockup/src/lib.rs [hunk 9]
```diff
@@ -926,53 +934,65 @@ mod tests {
...
+            to_nanos(YEAR).into(),
+            None,
+            TransfersInformation::TransfersEnabled {
+                transfers_timestamp: to_ts(GENESIS_TIME_IN_DAYS).into(),
+            },
+            VestingScheduleOrHash::VestingSchedule(vesting_schedule.clone()),
+            None,
+            AccountId::from("whitelist"),
+            Some(account_foundation()),
+        );
```
Comment (evgenykuzyakov) [inline lockup/src/lib.rs:958]: Can you update `new_contract` with whatever changes needed to avoid initializing the contract here?

File: lockup/src/types.rs [hunk 10]
```diff
@@ -45,7 +48,7 @@ pub struct LockupInformation {
 }
 
 /// Contains information about the transfers. Whether transfers are enabled or disabled.
-#[derive(BorshDeserialize, BorshSerialize, Deserialize, Serialize)]
+#[derive(BorshDeserialize, BorshSerialize, Deserialize, Serialize, Debug)]
```
Comment (evgenykuzyakov) [inline lockup/src/types.rs:51]: It seems `Debug` increases the size of the contract a lot. See +57Kb

Comment (evgenykuzyakov) [reply lockup/src/types.rs:51]: Apparently `Debug` is not the largest offender here. It's serde with tag, in particular this line:
```rust
#[serde(tag = "type", content = "value")]
```

It adds around 30Kb

**Comment by evgenykuzyakov** [comment]
Assigned to @ilblackdragon for review as well.

File: lockup/src/lib.rs (hunk 8 above)
Comment (evgenykuzyakov) [reply lockup/src/lib.rs:868]: It looks like I pasted them back above

File: lockup/src/types.rs [hunk 11]
```diff
@@ -165,3 +184,13 @@ pub struct TerminationInformation {
 /// The result of the transfer poll.
 /// Contains The timestamp when the proposal was voted in.
 pub type PollResult = Option<WrappedTimestamp>;
+
+pub fn hash_vesting_schedule(vesting_schedule: &VestingSchedule, salt: &[u8]) -> Hash {
```
Comment (nearmax) [inline lockup/src/types.rs:188]: I would suggest defining `VestingScheduleWithSalt` structure and serialize it as borsh.

File: lockup/src/internal.rs [hunk 12]
```diff
@@ -30,12 +30,19 @@ impl LockupContract {
...
-            env::panic(b"There is no vesting in progress");
-        }
+    pub fn assert_vesting(&self, vesting_schedule: &VestingSchedule, salt: &[u8]) {
+        match &self.vesting_information {
+            VestingInformation::VestingHash(hash) => assert_eq!(
+                &hash_vesting_schedule(vesting_schedule, salt),
+                &hash.0,
+                "Presented vesting doesn't match or vesting was terminated"
+            ),
+            VestingInformation::VestingSchedule(vs) => assert_eq!(
```
Comment (nearmax) [inline lockup/src/internal.rs:40]: It seems like we support a scenario when vesting schedule is saved directly in the contract but then foundation still needs to provide it to call contract methods. It is just a weird UX. We could keep it to make things simple, but we should explain it in the comment, otherwise people would think there is a deeper meaning to this.

File: lockup/src/types.rs (hunk 10 above)
Comment (nearmax) [reply lockup/src/types.rs:51]: FYI @mikedotexe @chadoh ^

File: lockup/src/lib.rs [hunk 13]
```diff
@@ -222,12 +212,23 @@ impl LockupContract {
...
+                    "Foundation account can't be added without vesting schedule"
+                );
+                VestingInformation::None
+            }
+            Some(VestingScheduleOrHash::VestingHash(hash)) => VestingInformation::VestingHash(hash),
+            Some(VestingScheduleOrHash::VestingSchedule(vs)) => {
+                VestingInformation::VestingSchedule(vs)
+            }
         };
+        assert!(
```
Comment (nearmax) [inline lockup/src/lib.rs:228]: This check looks irrelevant to adding vesting hash. Did we miss a bug in the past?

File: lockup/src/internal.rs (hunk 12 above)
Comment (evgenykuzyakov) [reply lockup/src/internal.rs:40]: Yep. Good point. Will update `terminate_vesting` method to take optional `VestingScheduleWithSalt`.

File: lockup/src/lib.rs (hunk 13 above)
Comment (evgenykuzyakov) [reply lockup/src/lib.rs:228]: Not really. Originally, I wrote the vesting schedule to be used for linear release without realizing we need both linear release and vesting schedule. So it was possible to have vesting schedule without foundation account in the past.

File: lockup/src/types.rs (hunk 11 above)
Comment (evgenykuzyakov) [reply lockup/src/types.rs]: done

**Review by bowenwang1996 — APPROVED** [review]

**Review by evgenykuzyakov — COMMENTED** [review]
I'm going to approve it. @ilblackdragon once you review it, feel free to merge

**Review by evgenykuzyakov — APPROVED** [review]

**Review by vgrichina — APPROVED** [review]
